cpp/daal/src/algorithms/kernel_function/polynomial/kernel_function_polynomial_dense_default_impl.i

```diff
@@ -0,0 +1,243 @@
+services::Status KernelImplPolynomial<defaultDense, algorithmFPType, cpu>::computeInternalVectorVector(const NumericTable * a1,
+                                                                                                       const NumericTable * a2, NumericTable * r,
+                                                                                                       const KernelParameter * par)
+{
+    if (par->kernelType != KernelType::linear)
+    {
+        return services::ErrorMethodNotImplemented;
+    }
+
+    //prepareData
+    const size_t nFeatures = a1->getNumberOfColumns();
+
+    ReadRows<algorithmFPType, cpu> mtA1(*const_cast<NumericTable *>(a1), par->rowIndexX, 1);
+    DAAL_CHECK_BLOCK_STATUS(mtA1);
+    const algorithmFPType * dataA1 = mtA1.get();
+
+    ReadRows<algorithmFPType, cpu> mtA2(*const_cast<NumericTable *>(a2), par->rowIndexY, 1);
+    DAAL_CHECK_BLOCK_STATUS(mtA2);
+    const algorithmFPType * dataA2 = mtA2.get();
+
+    WriteOnlyRows<algorithmFPType, cpu> mtR(r, par->rowIndexResult, 1);
+    DAAL_CHECK_BLOCK_STATUS(mtR);
+    algorithmFPType * dataR = mtR.get();
+
+    //compute
+    dataR[0] = 0.0;
+    PRAGMA_IVDEP
+    PRAGMA_VECTOR_ALWAYS
+    for (size_t i = 0; i < nFeatures; i++)
+    {
+        dataR[0] += dataA1[i] * dataA2[i];
+    }
+    dataR[0] = dataR[0] * par->scale + par->shift;
+
+    return services::Status();
+}
+
+template <typename algorithmFPType, CpuType cpu>
+services::Status KernelImplPolynomial<defaultDense, algorithmFPType, cpu>::computeInternalMatrixVector(const NumericTable * a1,
+                                                                                                       const NumericTable * a2, NumericTable * r,
+                                                                                                       const KernelParameter * par)
+{
+    if (par->kernelType != KernelType::linear)
+    {
+        return services::ErrorMethodNotImplemented;
+    }
+
+    //prepareData
+    const size_t nVectors1 = a1->getNumberOfRows();
+    const size_t nFeatures = a1->getNumberOfColumns();
+
+    ReadRows<algorithmFPType, cpu> mtA1(*const_cast<NumericTable *>(a1), 0, nVectors1);
+    DAAL_CHECK_BLOCK_STATUS(mtA1);
+    const algorithmFPType * dataA1 = mtA1.get();
+
+    ReadRows<algorithmFPType, cpu> mtA2(*const_cast<NumericTable *>(a2), par->rowIndexY, 1);
+    DAAL_CHECK_BLOCK_STATUS(mtA2);
+    const algorithmFPType * dataA2 = mtA2.get();
+
+    WriteOnlyRows<algorithmFPType, cpu> mtR(r, 0, nVectors1);
+    DAAL_CHECK_BLOCK_STATUS(mtR);
+    algorithmFPType * dataR = mtR.get();
+
+    //compute
+    algorithmFPType k = (algorithmFPType)(par->scale);
+    algorithmFPType b = (algorithmFPType)(par->shift);
+    for (size_t i = 0; i < nVectors1; i++)
+    {
+        dataR[i] = 0.0;
+        PRAGMA_IVDEP
+        PRAGMA_VECTOR_ALWAYS
+        for (size_t j = 0; j < nFeatures; j++)
+        {
+            dataR[i] += dataA1[i * nFeatures + j] * dataA2[j];
```

Comment:
gemv?

Also, it should solve a potential issue that your implementation is not threaded.

Reply:
I would leave the implementation as a reference, because this is a code branch that is currently not used anywhere.

Reply:
Simpler code?)